presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBytesValues.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+import com.google.common.collect.ImmutableList;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@State(Scope.Thread)
+@OutputTimeUnit(MILLISECONDS)
+@Fork(2)
+@Warmup(iterations = 10, time = 1000, timeUnit = MILLISECONDS)
+@Measurement(iterations = 10, time = 1000, timeUnit = MILLISECONDS)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkBytesValues
+{
+    private final byte[] word = new byte[]{'h', 'h', 'h'};
+
+    @Benchmark
+    public int lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
```

Comment:
I don't think this benchmark is valid. In general you are not supposed to perform same operation in a loop like this, rather let JMH do that. Re-writing the benchmark to remove the loop shows same performance for these two operations:

```
    @Benchmark
    public boolean lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
    {
        return data.exclusiveFilter.testBytes(word, 0, 3);
    }

    @Benchmark
    public boolean lookupInMultiRange(BenchmarkBytesValues.BenchmarkData data)
    {
        return data.multiRangeFilter.testBytes(word, 0, 3);
    }

# Run complete. Total time: 00:01:23

Benchmark                                Mode  Cnt   Score    Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20  ≈ 10⁻⁵           ms/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20  ≈ 10⁻⁵           ms/op
```

Writing good benchmark is not a requirement for the PR. If you want to write it, I suggest to start by writing benchmark for the main use case, e.g. in list with multiple entries, then see when performance difference is for very large in lists (e.g. hundreds of values), then make sure that in-list with one entry or != filter doesn't regress. Here is where I would start:

```
@OutputTimeUnit(NANOSECONDS)

    // NOT IN ("apple", "grape", "orange")
    // "abc", "apple", "banana", "grape", "orange", "peach"
    private final byte[][] testWords = new byte[][] {"abc".getBytes(), "apple".getBytes(), "banana".getBytes(), "grape".getBytes(), "orange".getBytes(), "peach".getBytes()};

    @Benchmark
    public int lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
    {
        int hits = 0;
        for (int i = 0; i < testWords.length; i++) {
            if (data.exclusiveFilter.testBytes(testWords[i], 0, testWords[i].length)) {
                hits++;
            }
        }

        return hits;
    }

    @Benchmark
    public int lookupInMultiRange(BenchmarkBytesValues.BenchmarkData data)
    {
        int hits = 0;
        for (int i = 0; i < testWords.length; i++) {
            if (data.multiRangeFilter.testBytes(testWords[i], 0, testWords[i].length)) {
                hits++;
            }
        }

        return hits;
    }

    @State(Scope.Thread)
    public static class BenchmarkData
    {
        private TupleDomainFilter exclusiveFilter;
        private TupleDomainFilter multiRangeFilter;

        @Setup
        public void setup()
                throws Exception
        {
            exclusiveFilter = BytesValuesExclusive.of(new byte[][] {"apple".getBytes(), "grape".getBytes(), "orange".getBytes()}, false);
            multiRangeFilter = TupleDomainFilter.MultiRange.of(
                    ImmutableList.of(
                            BytesRange.of(null, true, "apple".getBytes(), true, false),
                            BytesRange.of("apple".getBytes(), true, "grape".getBytes(), true, false),
                            BytesRange.of("grape".getBytes(), true, "orange".getBytes(), true, false),
                            BytesRange.of("orange".getBytes(), true, null, true, false)),
                    false,
                    false);
        }
    }

Benchmark                                Mode  Cnt   Score   Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20  63.112 ± 0.811  ns/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20  93.394 ± 0.530  ns/op
```

Reply:
Thank you for the detailed suggestion, it's really helpful!